pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -0,0 +1,218 @@
+	mapping, err := gc.restMapper.RESTMapping(fqKind.GroupKind(), apiVersion)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get REST mapping for kind: %s, version: %s", kind, apiVersion)
+	}
+	glog.Infof("map kind %s, version %s to resource %s", kind, apiVersion, mapping.Resource)
+	resource := unversioned.APIResource{
+		Name:       mapping.Resource,
+		Namespaced: namespace == "",
+		Kind:       kind,
+	}
+	return &resource, nil
+}
+
+func (gc *GarbageCollector) deleteObject(typeAccessor meta.Type, accessor meta.Object) error {
+	fqKind := unversioned.FromAPIVersionAndKind(typeAccessor.GetAPIVersion(), typeAccessor.GetKind())
+	client, err := gc.clientPool.ClientForGroupVersion(fqKind.GroupVersion())
+	resource, err := gc.apiResource(typeAccessor.GetAPIVersion(), typeAccessor.GetKind(), accessor.GetNamespace())
+	if err != nil {
+		return err
+	}
+	uid := accessor.GetUID()
+	preconditions := v1.Preconditions{UID: &uid}
+	deleteOptions := v1.DeleteOptions{Preconditions: &preconditions}
+	return client.Resource(resource, accessor.GetNamespace()).Delete(accessor.GetName(), &deleteOptions)
+}
+
+func (gc *GarbageCollector) checkGarbage(storeKey StoreKey) error {
+	obj, exists, err := storeKey.Store.GetByKey(storeKey.Key)
+	if err != nil {
+		return fmt.Errorf("unable to retrieve object %s from store: %v", storeKey.Key, err)
+	}
+	if !exists {
+		return fmt.Errorf("object has been deleted %s", storeKey.Key)
+	}
+	accessor, err := meta.Accessor(obj)
+	if err != nil {
+		return fmt.Errorf("unable to access ObjectMeta of object %s: %v", storeKey.Key, err)
+	}
+	typeAccessor, err := meta.TypeAccessor(obj)
+	if err != nil {
+		return fmt.Errorf("unable to access TypeMeta of object %s: %v", storeKey.Key, err)
+	}
+	ownerReferences := accessor.GetOwnerReferences()
+	if len(ownerReferences) == 0 {
+		glog.Infof("object %s's doesn't have an owner, continue on next object", storeKey.Key)
+		return nil
+	}
+	ownerExists := false
+	for _, reference := range ownerReferences {
+		fqKind := unversioned.FromAPIVersionAndKind(reference.APIVersion, reference.Kind)
+		client, err := gc.clientPool.ClientForGroupVersion(fqKind.GroupVersion())
+		if err != nil {
+			return err
+		}
+		resource, err := gc.apiResource(reference.APIVersion, reference.Kind, reference.Namespace)
+		if err != nil {
+			return err
+		}
+		_, err = client.Resource(resource, reference.Namespace).Get(reference.Name)
+		if err == nil {
+			ownerExists = true
+			break
+		} else if errors.IsNotFound(err) {
+			glog.Infof("object %s's owner %s/%s, %s is not found", storeKey.Key, reference.APIVersion, reference.Kind, reference.Name)
+		} else {
+			return err
+		}
+	}
+	if !ownerExists {
+		return gc.deleteObject(typeAccessor, accessor)
+	}
+	glog.Infof("object %s's still have exsisting owner, continue on next object", storeKey.Key)
+	return nil
+}
+
+func (gc *GarbageCollector) Run(workers int, stopCh <-chan struct{}) {
+	// list
+	go wait.Until(gc.scanner, time.Second, stopCh)
```

Comment:
time.Second is wrong. I think it could be any value larger than the ResourceResyncTime.